### ISSUE TITLE
Update external editor documentation

### ISSUE THREAD

**Issue by Beep6581** [Beep6581/RawTherapee#5474]
In light of recent changes, this needs updating:
http://rawpedia.rawtherapee.com/Preferences#External_Editor

For Linux, Windows and macOS, provide two working examples as follows:
1. Pass the image to a program with a space in its filename (rename the executable for testing purposes as needed), e.g.
`C:\Program Files\Adobe\Photoshop\Adobe Photoshop.exe`
or
`/home/bob/programs/hdrmerge/hdr merge`.
1. A program and a parameter, the parameter should have an observable effect so that you can confirm that it was parsed correctly, e.g.
`/usr/bin/geeqie --remote`
or
`open -a "Adobe Photoshop CS6" --some-paremeter` (I don't know how to pass a parameter to a program in macOS, because its not documented).

**Comment by blj86** [comment]
I can help a little bit here with MacOS.  They made some useful tweaks that allow passing of arguments to application with the open command in 10.6.2+.  So any version prior it will not work.

1.  open -a hdrmerge

2. open -a "Adobe Photoshop CS6"

3. open -a "Adobe Photoshop CS6" --args <arguments here>

For some more reference here is the man page for the open command on 10.14.6

`
OPEN(1) 		  BSD General Commands Manual		       OPEN(1)

NAME
     open -- open files and directories

SYNOPSIS
     open [-e] [-t] [-f] [-F] [-W] [-R] [-n] [-g] [-j] [-h] [-s sdk]
	  [-b bundle_identifier] [-a application] file ... [--args arg1 ...]

DESCRIPTION
     The open command opens a file (or a directory or URL), just as if you had
     double-clicked the file's icon. If no application name is specified, the
     default application as determined via LaunchServices is used to open the
     specified files.

     If the file is in the form of a URL, the file will be opened as a URL.

     You can specify one or more file names (or pathnames), which are inter-
     preted relative to the shell or Terminal window's current working direc-
     tory. For example, the following command would open all Word files in the
     current working directory:

     open *.doc

     Opened applications inherit environment variables just as if you had
     launched the application directly through its full path.  This behavior
     was also present in Tiger.

     The options are as follows:

     -a application
	 Specifies the application to use for opening the file

     -b bundle_indentifier
	 Specifies the bundle identifier for the application to use when open-
	 ing the file

     -e  Causes the file to be opened with /Applications/TextEdit

     -t  Causes the file to be opened with the default text editor, as deter-
	 mined via LaunchServices

     -f  Reads input from standard input and opens the results in the default
	 text editor.  End input by sending EOF character (type Control-D).
	 Also useful for piping output to open and having it open in the
	 default text editor.

     -F  Opens the application "fresh," that is, without restoring windows.
	 Saved persistent state is lost, except for Untitled documents.

     -W  Causes open to wait until the applications it opens (or that were
	 already open) have exited.  Use with the -n flag to allow open to
	 function as an appropriate app for the $EDITOR environment variable.

     -R  Reveals the file(s) in the Finder instead of opening them.

     -n  Open a new instance of the application(s) even if one is already run-
	 ning.

     -g  Do not bring the application to the foreground.

     -j  Launches the app hidden.

     -h  Searches header locations for a header whose name matches the given
	 string and then opens it.  Pass a full header name (such as NSView.h)
	 for increased performance.

     -s  For -h, partial or full SDK name to use; if supplied, only SDKs whose
	 names contain the argument value are searched. Otherwise the highest
	 versioned SDK in each platform is used.

     --args
	 All remaining arguments are passed to the opened application in the
	 argv parameter to main().  These arguments are not opened or inter-
	 preted by the open tool.

EXAMPLES
     "open '/Volumes/Macintosh HD/foo.txt'" opens the document in the default
     application for its type (as determined by LaunchServices).

     "open '/Volumes/Macintosh HD/Applications/'" opens that directory in the
     Finder.

     "open -a /Applications/TextEdit.app '/Volumes/Macintosh HD/foo.txt'"
     opens the document in the application specified (in this case, TextEdit).

     "open -b com.apple.TextEdit '/Volumes/Macintosh HD/foo.txt'" opens the
     document in the application specified (in this case, TextEdit).

     "open -e '/Volumes/Macintosh HD/foo.txt'" opens the document in TextEdit.

     "ls | open -f" writes the output of the 'ls' command to a file in /tmp
     and opens the file in the default text editor (as determined by Launch-
     Services).

     "open http://www.apple.com/" opens the URL in the default browser.

     "open 'file://localhost/Volumes/Macintosh HD/foo.txt'" opens the document
     in the default application for its type (as determined by LaunchSer-
     vices).

     "open 'file://localhost/Volumes/Macintosh HD/Applications/'" opens that
     directory in the Finder.

     "open -h NSView" lists headers whose names contain NSView and allows you
     to choose which ones to open.

     "open -h NSView.h" immediately opens NSView.h.

     "open -h NSView -s OSX10.12" lists headers whose names contain NSView in
     the MacOSX 10.12 SDK and allows you to choose which ones to open.

HISTORY
     First appeared in NextStep.

macOS				April 14, 2017				 macOS
`

**Comment by Beep6581** [comment]
Hi @blj86 , thank you for helping. Which version of RawTherapee did you test?

**Comment by blj86** [comment]
@Beep6581

I have done no specific tests as of this moment.  Just passed some information as I stumbled upon this while checking to see if their was info about a crash fix for 5.7 on MacOS.

If you have a specific version you need tests for let me know and I will look into figuring out a build setup.  I am loving RT so if I can help a little with what little skills I have I am willing.

**Comment by Beep6581** [comment]
@blj86 the point here is to test whether the three forms of invocation work correctly when used in RawTherapee via the "Open with external editor" button and to document that accordingly. To that end, testing should be done using `5.7-121-g`741bfed52 or newer.

**Comment by blj86** [comment]
@Beep6581

Ok sounds good will try to get a local build put together later tonight after work and run some tests asap.  Best way I can give back as I am not good enough with coding to contribute that way.

**Comment by blj86** [comment]
@Beep6581 
Update ran some tests.  Had some issues with getting the bundle I built to run but I was able to run the build by executing the built file instead of the bundle.

I used `5.7-122-gbd3c8a206`

Yes it seems the full bundle path is needed for the application will not work otherwise.

1. `open -a /Applications/GIMP-2.10.app`  = Works
2. `open -a "/Applications/GIMP-2.10.app"` = Works
(Spaces Test this is more of just to see if it works.  this application can't open images but it still launches properly in this test it is the only app I have currently with spaces in the name)
3. `open -a /Applications/Photo\ Booth.app` = Works
4. `open -a "/Applicatoins/Photo Booth.app"` = Works
5. `open -a /Applications/GIMP-2.10.app --args -a` = FAILS
6. `open -a "/Applications/GIMP-2.10.app" --args -a` = FAILS

So -a in gimp opens the image as a new image tab in the editor.  Through some manual command line testing the arguments do work but the formatting is very important.

`open -a /Applications/GIMP-2.10.app <file> --args -a`
`open -a "/Applications/GIMP-2.10.app" <file> --args -a`

In other words the list of files being passed to gimp on MacOS must come before the arguments you are passing to GIMP.  So this particular edge case on MacOS may need some special code written to reformat the external command to ensure the --args comes after the actual image being passed through.

Hope this is helpful.

**Comment by Lawrence37** [comment]
The documentation also need to explain the output directory section added in #6232. It should be clear that images saved to the temporary directory may be automatically removed by the operating system, typically on reboot for Linux and macOS (see https://discuss.pixls.us/t/external-editor-adobe-photoshop-installation-directory/27994/13). The page on [saving images](http://rawpedia.rawtherapee.com/Saving_Images) implies that sending an image to an external editor is enough to save it. There should be a sentence to say that images aren't really saved until it is saved using the external editor or if the external editor output directory is not a temporary directory.
